### PR TITLE
bugfix for generics

### PR DESCRIPTION
**Pull request by isc30**
- [x] Parser
- [x] Passing `in` generics
- [x] Generic implementations (uses an anonymous symbol that inherits from the constraints for code generation)
- [ ] Validators for the caller (compiler error if constraints aren't matched)

Not supported:
- `typeof(T)` due to JavaScript runtime limitations (they have the same issue in TypeScript)

### CODE REVIEW

File: src/DSharp.Compiler/Compiler/ExpressionBuilder.cs [hunk 1]
```diff
@@ -1825,6 +1820,10 @@ private Expression ProcessOpenParenExpressionNode(BinaryExpressionNode node)
                 }
 
                 methodExpression = new MethodExpression(memberExpression.ObjectReference, method);
+
+                // iterate over the parameters. If the parameter is generic, then get args[i].EvaluatedType
```
Comment (isc30) [inline src/DSharp.Compiler/Compiler/ExpressionBuilder.cs]: this is what needs to be done in order to support `typeof(T)` and `T` as return type

File: src/DSharp.Compiler.Tests/Source/Expression/Generics2/Baseline.txt [hunk 2]
```diff
@@ -0,0 +1,49 @@
+"use strict";
+
+define('g_b4bdbc8e390b4f5c8e02af2037e648c6', ['ss'], function(ss) {
```
Comment (fred-perkins) [inline src/DSharp.Compiler.Tests/Source/Expression/Generics2/Baseline.txt]: You'll want a ScriptAssemlyAttribute here to stop it from using a generated name. Otherwise it'll fail between runs - Something I want to fix!

File: src/DSharp.Compiler/Compiler/MetadataBuilder.cs [hunk 3]
```diff
@@ -801,6 +801,44 @@ private MethodSymbol BuildMethod(MethodDeclarationNode methodNode, TypeSymbol ty
                 if (returnType != null)
                 {
                     method = new MethodSymbol(methodNode.Name, typeSymbol, returnType, methodNode.IsExensionMethod);
+
+                    if (methodNode.TypeParameters.Any())
+                    {
+                        List<GenericParameterSymbol> genericArguments = new List<GenericParameterSymbol>();
+
+                        int i = 0;
```
Comment (fred-perkins) [inline src/DSharp.Compiler/Compiler/MetadataBuilder.cs]: Why not just do a for loop?

File: src/DSharp.Compiler/ScriptModel/Symbols/GenericParameterSymbol.cs [hunk 4]
```diff
@@ -1,11 +1,11 @@
-// GenericParameterSymbol.cs
+﻿// GenericParameterSymbol.cs
 // Script#/Core/Compiler
 // This source code is subject to terms and conditions of the Apache License, Version 2.0.
 //
 
 namespace DSharp.Compiler.ScriptModel.Symbols
 {
-    internal sealed class GenericParameterSymbol : TypeSymbol
+    internal sealed class GenericParameterSymbol : ClassSymbol
```
Comment (fred-perkins) [inline src/DSharp.Compiler/ScriptModel/Symbols/GenericParameterSymbol.cs]: Why the change to class symbols only? Can't remember if Structs go under class - Have a feeling they do, but unsure :D

File: src/DSharp.Compiler/Compiler/ExpressionBuilder.cs (hunk 1 above)
Comment (fred-perkins) [reply src/DSharp.Compiler/Compiler/ExpressionBuilder.cs]: We then also need to include those type parameters in the method parameters when invoked by callers, same with if a class is constructed, we need a mechanism to pass those generic parameters in, as the type check is a runtime check

File: src/DSharp.Compiler/ScriptModel/Symbols/GenericParameterSymbol.cs (hunk 4 above)
Comment (isc30) [reply src/DSharp.Compiler/ScriptModel/Symbols/GenericParameterSymbol.cs]: yeah, a emulated generic type is a class in the sense that it can have a single base class [0..1] and [0..N] interfaces

File: src/DSharp.Compiler/Compiler/MetadataBuilder.cs [hunk 5]
```diff
@@ -801,6 +801,45 @@ private MethodSymbol BuildMethod(MethodDeclarationNode methodNode, TypeSymbol ty
                 if (returnType != null)
                 {
                     method = new MethodSymbol(methodNode.Name, typeSymbol, returnType, methodNode.IsExensionMethod);
+
+                    if (methodNode.TypeParameters.Any())
+                    {
+                        List<GenericParameterSymbol> genericArguments = new List<GenericParameterSymbol>();
```
Comment (fred-perkins) [inline src/DSharp.Compiler/Compiler/MetadataBuilder.cs]: Can we put this into a new method if possible?

File: src/DSharp.Compiler.Tests/Source/Expression/ScriptIgnore/Baseline.txt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+define('test', ['ss'], function(ss) {
+  var $global = this;
+  // ExpressionTests.Program
+
+  function Program() {
+  }
+  var Program$ = {
+    main: function() {
+      var e1 = ss.enumerate(this);
+      var e2 = ss.enumerate((this));
```
Comment (fred-perkins) [inline src/DSharp.Compiler.Tests/Source/Expression/ScriptIgnore/Baseline.txt]: Interesting that it double braces here.

Comment (isc30) [reply src/DSharp.Compiler.Tests/Source/Expression/ScriptIgnore/Baseline.txt]: it's because of the parenthesis surrounding the cast to be able to call the method `((X)b).a()` => `(b).a()` => `ss.a((b))`

File: src/DSharp.Compiler/Compiler/MetadataBuilder.cs [hunk 7]
```diff
@@ -1276,8 +1276,16 @@ private void BuildTypeInheritance(ClassSymbol classSymbol)
 
                 foreach (NameNode node in customTypeNode.BaseTypes)
                 {
+                    string nodeName = node.Name;
+
+                    if (node is GenericNameNode genericNameNode)
+                    {
+                        nodeName += $"`{genericNameNode.TypeArguments.Count}";
```
Comment (fred-perkins) [inline src/DSharp.Compiler/Compiler/MetadataBuilder.cs]: Could we not push this down into the GenericNameNode and have a new property like `GenericName` to expose this?

Comment (isc30) [reply src/DSharp.Compiler/Compiler/MetadataBuilder.cs]: lets leave this for the future

Comment (fred-perkins) [reply src/DSharp.Compiler/Compiler/MetadataBuilder.cs]: Sure, It's not urgent. I can always refactor in any generics work.
